derive-eip712/src/lib.rs

```diff
@@ -0,0 +1,399 @@
+                    .path
+                    .segments
+                    .first()
+                    .map(|s| s.ident.to_string())
+                    .expect("argument must have an identity");
+
+                field_type.push_str(&arg_identity);
+            }
+        }
+        field_type.push('>');
+    }
+
+    parse_field_type(field_type)
+}
+
+// Return HashMap of the field name and the field type;
+fn parse_fields(ast: &syn::DeriveInput) -> HashMap<String, String> {
+    let mut parsed_fields = HashMap::new();
+
+    let data = match &ast.data {
+        syn::Data::Struct(s) => s,
+        _ => {
+            panic!("Eip712 can only be derived for a struct")
+        }
+    };
+
+    let named_fields = match &data.fields {
+        syn::Fields::Named(name) => name,
+        _ => {
+            panic!("unnamed fields are not supported")
+        }
+    };
+
+    named_fields.named.iter().for_each(|f| {
+        let field_name = f
+            .ident
+            .clone()
+            .expect("field must be named")
+            .to_string()
+            .to_case(Case::Camel);
+
+        let field_type = parse_field(f);
+
+        parsed_fields.insert(field_name, field_type);
+    });
+
+    parsed_fields
+}
+
+// Convert hash map of field names and types into a type hash corresponding to enc types;
+fn make_type_hash(primary_type: String, fields: &HashMap<String, String>) -> String {
+    let parameters = fields
+        .iter()
+        .map(|(k, v)| format!("{} {}", v, k))
+        .collect::<Vec<String>>()
+        .join(",");
+
+    let sig = format!("{}({})", primary_type, parameters);
+
+    hex::encode(keccak256(sig))
+}
+
+// Main implementation macro, used to compute static values and define
+// method for encoding the final eip712 payload;
+fn impl_eip_712_macro(ast: &syn::DeriveInput) -> TokenStream {
+    // Primary type should match the type in the ethereum verifying contract;
+    let primary_type = &ast.ident;
+
+    // Computer domain separator
+    let domain_attributes: Eip712Domain = Eip712Domain::from(ast);
+    let domain_separator = domain_attributes.separator();
+    let domain_type_hash = hex::encode(eip712_domain_type_hash());
+
+    // Must parse the AST at compile time.
+    let parsed_fields = parse_fields(ast);
+
+    // JSON Stringify the field names and types to pass into the
+    // derived encode_eip712() method as a static string;
+    // the AST of the struct is not available at runtime, so this is
+    // a work around for passing in the struct fields;
+    let fields: String = serde_json::to_string(&parsed_fields)
+        .expect("failed to serialize parsed fields into JSON string");
+
+    // Compute the type hash for the derived struct using the parsed fields from above;
+    let type_hash = make_type_hash(primary_type.clone().to_string(), &parsed_fields);
+
+    let implementation = quote! {
+        #[derive(Debug, thiserror::Error)]
+        pub enum Eip712Error {
```

Comment:
Move to `ethers_core::transaction::eip712`. Multiple derivations while erroneously call multiple declarations.

Reply:
moved